python/cuda_parallel/cuda/parallel/experimental/_iterators.py

```diff
@@ -55,12 +35,76 @@ def _ctypes_type_given_numba_type(ntype):
     return mapping[ntype]
 
 
-CompileResult = collections.namedtuple('CompileResult', ['ltoir', 'result_type'])
+@lru_cache(maxsize=256)  # TODO: what's a reasonable value?
+def cached_compile(func, sig, abi_name=None, **kwargs):
+    return numba.cuda.compile(func, sig, abi_info={"abi_name": abi_name}, **kwargs)
+
+
+class IteratorBase:
+    """
+    An Iterator is a wrapper around a pointer, and must define the following:
+
+    - a `state` property that returns a `ctypes.c_void_p` object, representing
+      a pointer to some data.
+    - an `advance` (static) method that receives the state pointer and performs
+      an action that advances the pointer by the offset `distance`
+      (returns nothing).
+    - a `dereference` (static) method that dereferences the state pointer
+      and returns a value.
+    """
+
+    def __init__(self, numba_type, value_type, abi_name):
+        """
+        Parameters
+        ----------
+        numba_type
+          A numba type that specifies how to interpret the state pointer.
+        value_type
+          The numba type of the value returned by the dereference operation.
+        abi_name
+          A unique identifier that will determine the abi_names for the
+          advance and dereference operations.
+        """
+        self.numba_type = numba_type
+        self.value_type = value_type
+        self.abi_name = abi_name
+
+    # TODO: should we cache this? Current docs environment doesn't allow
+    # using Python > 3.7. We could use a hand-rolled cached_property if
+    # needed.
+    @property
+    def ltoirs(self):
+        advance_abi_name = self.abi_name + "_advance"
+        deref_abi_name = self.abi_name + "_dereference"
+        advance_ltoir, _ = cached_compile(
+            self.__class__.advance,
+            (
+                self.numba_type,
+                numba.types.uint64,  # distance type
+            ),
+            output="ltoir",
+            abi_name=advance_abi_name,
+        )
+
+        deref_ltoir, _ = cached_compile(
+            self.__class__.dereference,
+            (self.numba_type,),
+            output="ltoir",
+            abi_name=deref_abi_name,
+        )
+        return {advance_abi_name: advance_ltoir, deref_abi_name: deref_ltoir}
```

Comment:
It's safer to simply return a tuple of tuples here, then there is no question about insertion order.

Nicest would be a `namedtuple` or `dataclass`, with `abi_name`, `ltoir` as members. But since there is only one caller (IIUC) a plain tuple of tuples would seem fine.

Reply:
There's no risk here from using a `dict`. Python dicts guarantee maintaining insertion order since Python 3.7. From the Python [docs](https://docs.python.org/3/library/collections.html#ordereddict-objects):

> [...] the built-in [dict](https://docs.python.org/3/library/stdtypes.html#dict) class gained the ability to remember insertion order (this new behavior became guaranteed in Python 3.7):

A `namedtuple` may be a little nicer but a `dict` serves fine, especially for an internal function.